src/System.Buffers/src/System/Buffers/BufferBucket.cs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace System.Buffers
+{
+    internal sealed class BufferBucket
+    {
+        private volatile int _index;
+        private byte[][] _data;
+        private int _bufferSize;
+        private SpinLock _lock;
+
+        internal BufferBucket(int bufferSize, int numberOfBuffers)
+        {
+            _index = 0;
+            _bufferSize = bufferSize;
+            _lock = new SpinLock();
+            _data = new byte[numberOfBuffers][];
+            for (int i = 0; i < numberOfBuffers; i++)
+            {
+                _data[i] = new byte[bufferSize];
```

Comment:
Seems a bit generous to allocate all buffers upfront. We may need just one or two.
Perhaps these could be lazy allocated in Rent when actually needed?